test: unit testing for lib and such

favna opened the pull request:
This comes with some pretty major refactors, so this PR will get a SEM: Major label. However these refactors are required to have a stable, easy to work with testing environment that is also scalable to the varied developers that work on Skyra. So without further ado the changes made:

1. Ensured that en-US and es-ES have all keys between them and all in the same sorting order. This was possible to track down thanks to the unit test!
1. Implemented TypeScript composite builds and project references. I tried to avoid it but it turned out to be a must-have requirement because otherwise ts-jest wouldn't pick up on things like module augmentation.
    - Updated all the scripts in package.json
    - Updated the tsc workflow for this. I can atest this works because I use it myself (and iCrawl and I had to go through various iterations because what even is testing)
    - Enabbled outputting declaration files from source. This is a requirement that comes with composite builds.
2. As side effect of that, `config.ts` and `config.example.ts` now reside in `src`. This is because they need to be in the same directory as the composite "src" project
3. Added the `__tests__` and `__mocks__` directories, which are default directory names for Jest.
4. Updated linting and building workflows to remove some old params for the respective actions.
5. Setup Jest and covered language files to the most extend that can be done.

resolves #546

Note: Merging this will likely cause merge conflicts to most other PR's, if not all!

favna committed:
test: setup unit testing

This comes with some pretty major refactors, so this PR will get a SEM: Major label. However these refactors are required to have a stable, easy to work with testing environment that is also scalable to the varied developers that work on Skyra. So without further ado the changes made:

0. Ensured that en-US and es-ES have all keys between them and all in the same sorting order. This was possible to track down thanks to the unit test!
1. Implemented TypeScript composite builds and project references. I tried to avoid it but it turned out to be a must-have requirement because otherwise ts-jest wouldn't pick up on things like module augmentation.
  - Updated all the scripts in package.json
  - Updated the tsc workflow for this. I can atest this works because I use it myself (and iCrawl and I had to go through various iterations because what even is testing)
  - Enabbled outputting declaration files from source. This is a requirement that comes with composite builds.
2. As side effect of that, `config.ts` and `config.example.ts` now reside in `src`. This is because they need to be in the same directory as the composite "src" project
3. Added the __tests__ and __mocks__ directories, which are default directory names for Jest.
  - Took most of the mocks from @klasa/settings-gateway, adding only the MockLanguageStore to it.
4. Updated linting and building workflows to remove some old params for the respective actions.
5. Setup Jest and ran the tests to make them work locally at least

Signed-off-by: Jeroen Claassens <support@favware.tech>

favna committed:
style: fix a bunch of linting errors

Signed-off-by: Jeroen Claassens <support@favware.tech>

favna committed:
build: fixed build and run process

Signed-off-by: Jeroen Claassens <support@favware.tech>

favna committed:
ci: fix build workflow

Signed-off-by: Jeroen Claassens <support@favware.tech>

kyranet commented on src/languages/es-ES.ts:
```diff
@@ -2519,12 +2547,12 @@ export default class extends Language {
+			HOW_MUCH: 'cuanto',
+			HOW_MANY: 'cuantos',
+			WHY: 'por qué',
+			WHO: 'quien'
```
... why? Those changes are wrong

favna replied:
They now match the same keys for English where they were literal strings rather than Regex expressions as well. Either one of them had to have been wrong. I assumed it was the Spanish.

kyranet replied:
That object supports either strings or RegExp, please revert that change

kyranet commented on src/lib/queries/postgres.ts:
```diff
@@ -64,6 +64,7 @@ export class PostgresCommonQuery implements CommonQuery {
 	}
 
+	// eslint-disable-next-line @typescript-eslint/require-await
 	public async deleteMemberSettings(guildID: string, userID: string) {
```
Just remove the `async`

kyranet commented on src/lib/queries/postgres.ts:
```diff
@@ -137,6 +138,7 @@ export class PostgresCommonQuery implements CommonQuery {
 	}
 
+	// eslint-disable-next-line @typescript-eslint/require-await
 	public async purgeTwitchStreamGuildSubscriptions(guildID: string) {
```
Ditto

favna committed:
ci: fix test workflow

kyranet commented on __tests__/i18n.test.ts:
```diff
@@ -0,0 +1,34 @@
+		return [...res, ...i18nReducer(obj[el], `${prefix}${el}.`)];
+	}
+	return [...res, prefix + el];
+}, []);
```
This looks a lot like Klasa's `objectToTuples`, except far more compute expensive (`@klasa/utils`'s `objectToTuples` iterates only once, this one does Object.keys, reduces, and concatenates, which are 3 iterations).

Maybe adapt the klasa code? https://github.com/dirigeants/utils/blob/master/src/lib/objectToTuples.ts

kyranet reviewed (changes requested):
I don't understand the need of those mocks, we already have the classes in `src`... And `@klasa/settings-gateway` exporting `Client` is a bug I'll have to fix, so don't rely on it.

Quantumlyy commented on __tests__/i18n.test.ts:
```diff
@@ -0,0 +1,34 @@
+	} else if (typeof obj[el] === 'object' && obj[el] !== null) {
+		return [...res, ...i18nReducer(obj[el], `${prefix}${el}.`)];
+	}
+	return [...res, prefix + el];
```
```suggestion
	return [...res, `${prefix}${el}`];
```

favna replied:
Adopting Kyra's feedback instead about the util function. Resolving this one.

Quantumlyy commented on package.json:
```diff
@@ -84,11 +85,13 @@
 	"_moduleAliases": {
-		"@utils": "dist/src/lib/util",
-		"@lib": "dist/src/lib",
+		"@utils": "dist/lib/util",
+		"@lib": "dist/lib",
```
Why was this needed? It seemed correct the way it was previously. Mind explaining so it is visible if any other reviews request it?

kyranet replied:
Because TSC, when suggesting those paths, it doesn't do in order of length, but in key insertion order.

By placing `@utils` before `@lib`, it'll suggest `@utils/util.ts` before `@lib/utils/util.ts`, and automatic import will take the first one.

favna replied:
Unless you meant that it was `dist/src/lib` before, in which case it's because with this change there is no `src` directory in `dist` anymore as the `rootDir` TSC config option has been set.

Quantumlyy replied:
That is what it was directed it. So I ask why was dist/src/ moved to just dist/

favna committed:
chore: address feedback

favna committed:
test: address more feedback

- Use a different strategy for object to tuples
  - Yay this also revealed that apparently the previous one was flawed, had to add 3 more keys to es-ES
- Removed a lot of mocks that were not even needed

Signed-off-by: Jeroen Claassens <support@favware.tech>

favna committed:
revert 8ball es-es change

Rexogamer commented on package.json:
```diff
@@ -2,23 +2,23 @@
 	"version": "5.2.3",
 	"description": "Multipurpose Discord Bot built on Klasa",
-	"main": "dist/src/Skyra.js",
+	"main": "dist/Skyra.js",
```
Is there such a file? Thought everything was in TS.

kyranet replied:
Yeah, after compiling Skyra from TS to JS, `dist/Skyra.js` exists

Rexogamer commented on .vscode/launch.json:
```diff
@@ -1,35 +1,35 @@
+			"request": "launch",
+			"name": "Main - Windows",
+			"args": [
+				"${workspaceFolder}\\dist\\src\\Skyra.js"
```
Same thing here

kyranet replied:
```suggestion
				"${workspaceFolder}\\dist\\Skyra.js"
```

favna replied:
gdi I searched for `src/Skyra`, forgetting about the `src\\Skyra` lol. Thanks for catching it guys.

favna committed:
Update .vscode/launch.json

Co-Authored-By: Antonio Román <kyradiscord@gmail.com>

favna commented on __tests__/i18n.test.ts:
```diff
@@ -0,0 +1,37 @@
+ * @param value The object to convert
+ * @param prefix The prefix for the key
+ */
+export default function objectToTuples<T = Record<string, unknown>>(original: T, prefix = ''): string[] {
```
```suggestion
function objectToTuples<T = Record<string, unknown>>(original: T, prefix = ''): string[] {
```

Woops, that was a copypasta mistake

favna committed:
Update __tests__/i18n.test.ts

favna committed:
test(i18n): cover remaining functions

favna committed:
ci: fix node-canvas issues in gh action

favna committed:
test: cleanup

favna committed:
ci: cleanup test flow

favna committed:
test: cleanup i18n test

favna committed:
docs: update readme

kyranet commented on .github/CONTRIBUTING.md:
```diff
@@ -30,3 +30,6 @@ There are a number of guidelines considered when reviewing Pull Requests to be m
 - Everything should follow our ESLint rules as closely as possible, and should pass lint tests even if you must disable a rule for a single line.
 - Everything should follow [Discord Bot Best Practices](https://github.com/meew0/discord-bot-best-practices)
+
+
```
🧹
```suggestion
```

kyranet commented on .github/CONTRIBUTING.md:
```diff
@@ -30,3 +30,6 @@ There are a number of guidelines considered when reviewing Pull Requests to be m
 - Everything should follow [Discord Bot Best Practices](https://github.com/meew0/discord-bot-best-practices)
+
+
+[`config.example.ts`]:  /src/config.example.ts
```
```suggestion
[`config.example.ts`]: /src/config.example.ts
```

favna committed:
Apply suggestions from code review

Co-Authored-By: Antonio Román <kyradiscord@gmail.com>

cfanoulis commented on src/tsconfig.json:
```diff
@@ -0,0 +1,25 @@
+	"include": [
+		"./**/*"
+	]
+}
```
pls add a finishing eol

cfanoulis commented on tsconfig.eslint.json:
```diff
@@ -0,0 +1,10 @@
+		"./__mocks__/**/*",
+		"jest.config.js"
+	]
+}
```
pls add a finishing eol

cfanoulis reviewed (changes requested):
general comments

favna committed:
chore: formatting (enkiel's feedback)

favna committed:
docs: fix indent in readme
